### PR TITLE
Fix deadlock between 'podman ps' and 'container inspect' commands

### PR DESCRIPTION
**Pull request by tyler92**
Fixes: #16326

[NO NEW TESTS NEEDED]

Signed-off-by: Mikhail Khachayants <tyler92@inbox.ru>

#### Does this PR introduce a user-facing change?

```release-note
Fix deadlock between 'podman ps' and 'container inspect' commands
```

### CODE REVIEW

File: libpod/pod_api.go [hunk 1]
```diff
@@ -581,20 +581,22 @@ func (p *Pod) Status() (map[string]define.ContainerStatus, error) {
 }
 
 func containerStatusFromContainers(allCtrs []*Container) (map[string]define.ContainerStatus, error) {
-	// We need to lock all the containers
```
Comment (tyler92) [inline libpod/pod_api.go:584]: Of course, I saw this comment and my assumption is this is not correct now. Please let me know if I am wrong.

**Comment by baude** [comment]
/approve

i'm looking deeper on this r/n. i dont know about that assumption but there is an easy enough way to find ou8t

**Comment by baude** [comment]
@mheon i dont see evidence that this is no longer true unless i missed some change to container and pod locking.  What *might* make more sense would be to iterate the containers, grab a lock, get the container information, release the lock?  One loop vs two ?

**Comment by mheon** [comment]
Code looks fine as written, though personally would have avoided the anonymous function. The core issue is similar to the reason we had to rewrite pod removal - holding more than one container lock simultaneously, without specific ordering, can cause deadlocks with other commands that want to lock, say, a container and its dependencies. We got rid of one instance of this (pod removal), this looks like another.

**Comment by mheon** [comment]
LGTM, for reference

**Comment by tyler92** [comment]
> personally would have avoided the anonymous function

Can rewrite with explicit Unlock calls and without an anonymous function

File: libpod/pod_api.go [hunk 2]
```diff
@@ -581,20 +581,22 @@ func (p *Pod) Status() (map[string]define.ContainerStatus, error) {
...
+			defer ctr.lock.Unlock()
+
+			if err := ctr.syncContainer(); err != nil {
+				return err
+			}
+
+			status[ctr.ID()] = ctr.state.State
+			return nil
+		}
+		if err := getStateFunc(); err != nil {
```
Comment (vrothberg) [inline libpod/pod_api.go:597]: We can call `(*Container).State()` instead.

**Comment by tyler92** [comment]
I've published my small pet-project tool for debugging such deadlocks (in case it's impossible to use IDE or GDB): https://github.com/tyler92/podman-locks. It would be great if someday it will help someone other than me.

(I hope it's allowed to attach such links here)

**Review by vrothberg — COMMENTED** [review]
LGTM, thanks!

**Comment by vrothberg** [comment]
> I've published my small pet-project tool for debugging such deadlocks (in case it's impossible to use IDE or GDB): https://github.com/tyler92/podman-locks. It would be great if someday it will help someone other than me.

Thanks for sharing!  I usually use the BPF tools from `bcc-tools`, for instance, /usr/share/bcc/tools/deadlock.

**Comment by tyler92** [comment]
>  I usually use the BPF tools from bcc-tools,

Thanks, I will take it into account! The main difference with my tool is that it can be launched _after_ the deadlock has occurred and the podman process has already started as usual without any external tools.

**Comment by vrothberg** [comment]
Thanks, @ashley-cui :)

**Pull request merged**